### ISSUE TITLE
Investigation of non-gradient based methods for tree-based methods

### ISSUE THREAD

**Issue by MechCoder** [scikit-optimize/scikit-optimize#250]
It may be possible to use non gradient-based techniques like `fmin_powell`. See (http://www.scipy-lectures.org/advanced/mathematical_optimization/index.html#a-shooting-method-the-powell-algorithm)

**Comment by MechCoder** [comment]
Will investigate more tomorrow / in the coming week.

**Comment by betatim** [comment]
Talked with @ibab yesterday and one idea that came up was if it was possible for additive tree ensembles (aka gradient boosting) to find the minimum by walking the leaves of each tree successively. Instead of having to actually run a minimizer.

**Comment by MechCoder** [comment]
For a Random Forest this almost corresponds to checking every point in the training data right?

**Comment by betatim** [comment]
So far I couldn't think of a shortcut for forests, so I think there you have to use an optimizer.

Having thought about gradient boosted trees a bit more I am now unsure that what I wrote in my previous comment would work. So I think for GBRT there are no shortcuts either.

For a single decision tree however we can cheat...

**Comment by glouppe** [comment]
I think what this boils down to is how to compactly construct the partition of the input space corresponding to the random forest, so that we can exhaustively iterate over the sub-regions. For a tree, the sub-regions are the T leaves composing that tree, that's easy. For two trees there is at most T1*T2 sub-regions, but in practice that should be a loose upper bound as most leaves will not overlap. So I still think this complexity can be cut down. (Then if we solve the case for two trees, we can proceed by induction for the whole forest.)
